<a href="https://colab.research.google.com/github/maryamhehe/Machine-Learning/blob/master/FoodModel.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install tensorflow
!pip install tensorflow-object-detection-api
!pip install --upgrade dataset-tools

In [ ]:
import tensorflow as tf
print(tf.config.list_physical_devices('GPU'))

2024-10-19 09:56:36.667465: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-10-19 09:56:36.689973: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-10-19 09:56:36.697386: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-10-19 09:56:36.714964: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-10-19 09:56:38.007077: W tensorflow/comp

[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]


I0000 00:00:1729331800.292502    1957 cuda_executor.cc:1015] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
I0000 00:00:1729331801.023960    1957 cuda_executor.cc:1015] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
I0000 00:00:1729331801.024348    1957 cuda_executor.cc:1015] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355


In [ ]:
import dataset_tools as dtools
dtools.download(dataset='Food Recognition 2022', dst_dir='/content/food_recognition-dataset')


Unpacking 'food-recognition-2022.tar': 100%|██████████| 87927/87927 [05:22<00:00, 272.55file/s] 


'/content/food_recognition-dataset/food-recognition-2022'

In [ ]:
import os
# List the contents of the unpacked dataset directory
dataset_dir = '/content/food_recognition-dataset/food-recognition-2022'
os.listdir(dataset_dir)

['LICENSE.md', 'test', 'README.md', 'validation', 'meta.json', 'training']

In [ ]:
training_dir = os.path.join(dataset_dir, 'training')
validation_dir = os.path.join(dataset_dir, 'validation')

print("Training files:", os.listdir(training_dir))
print("Validation files:", os.listdir(validation_dir))


Training files: ['img', 'ann']
Validation files: ['img', 'ann']


In [ ]:
training_img_dir = os.path.join(training_dir, 'img')
training_ann_dir = os.path.join(training_dir, 'ann')
validation_img_dir = os.path.join(validation_dir, 'img')
validation_ann_dir = os.path.join(validation_dir, 'ann')

print("Training Image files:", os.listdir(training_img_dir))
print("Training Annotation files:", os.listdir(training_ann_dir))
print("Validation Image files:", os.listdir(validation_img_dir))
print("Validation Annotation files:", os.listdir(validation_ann_dir))


Training Image files: ['149978.jpg', '056801.jpg', '081526.jpg', '032168.jpg', '154007.jpg', '059351.jpg', '087271.jpg', '137558.jpg', '180964.jpg', '182660.jpg', '104789.jpg', '136097.jpg', '085456.jpg', '030531.jpg', '028152.jpg', '083019.jpg', '163715.jpg', '032273.jpg', '046696.jpg', '123158.jpg', '009226.jpg', '144578.jpg', '029252.jpg', '062357.jpg', '152268.jpg', '098673.jpg', '063052.jpg', '039350.jpg', '154177.jpg', '127893.jpg', '064403.jpg', '068253.jpg', '039085.jpg', '098986.jpg', '049044.jpg', '163993.jpg', '020022.jpg', '008109.jpg', '016748.jpg', '160226.jpg', '180280.jpg', '124430.jpg', '182526.jpg', '166808.jpg', '157818.jpg', '083662.jpg', '022709.jpg', '052479.jpg', '010044.jpg', '171582.jpg', '140034.jpg', '098125.jpg', '125342.jpg', '047793.jpg', '072533.jpg', '128353.jpg', '174449.jpg', '127346.jpg', '060209.jpg', '039364.jpg', '128228.jpg', '169256.jpg', '121528.jpg', '160145.jpg', '035616.jpg', '155497.jpg', '055404.jpg', '135998.jpg', '177028.jpg', '094576.jpg

In [ ]:
import json

# Load a sample annotation file
sample_annotation_path = os.path.join(training_ann_dir, '163930.jpg.json')
with open(sample_annotation_path, 'r') as f:
    annotation_data = json.load(f)

print(json.dumps(annotation_data, indent=4))


{
    "description": "",
    "tags": [],
    "size": {
        "height": 796,
        "width": 796
    },
    "objects": [
        {
            "id": 1731344,
            "classId": 3441,
            "description": "",
            "geometryType": "polygon",
            "labelerLogin": "az@datasetninja.com",
            "createdAt": "2023-06-07T14:06:32.755Z",
            "updatedAt": "2023-06-07T14:06:32.755Z",
            "tags": [],
            "classTitle": "tea",
            "points": {
                "exterior": [
                    [
                        379,
                        463
                    ],
                    [
                        314,
                        460
                    ],
                    [
                        284,
                        452
                    ],
                    [
                        243,
                        435
                    ],
                    [
                        222,
              

description: An empty string here; typically used to describe the image or the annotation.
tags: An empty list; tags can be used for categorizing or adding metadata to the image or annotation.
size: This section provides the dimensions of the image.
height: The height of the image in pixels (796).
width: The width of the image in pixels (796).
objects: This key contains a list of detected objects in the image. Each object includes various details:

id: A unique identifier for the object (1731344).
classId: An identifier corresponding to the class of the object (3441). This ID usually refers to a specific category in your dataset (e.g., "tea").
description: An empty string, which could be used for additional information about the object.
geometryType: Specifies the type of geometry used for annotation. In this case, it's a "polygon," which indicates the object's shape is defined by a series of points.
labelerLogin: The identifier of the user who annotated the object (az@datasetninja.com).
createdAt: The timestamp indicating when the annotation was created.
updatedAt: The timestamp indicating when the annotation was last updated.
tags: An empty list, similar to the top-level tags; it can be used for additional categorization.
classTitle: The human-readable name of the object class ("tea").
points: This section contains the coordinates that define the shape of the object in the image.

exterior: A list of points that outline the outer shape of the object, represented as pairs of (x, y) coordinates. Each pair corresponds to a vertex in the polygon.
interior: An empty list; if the object had holes, the points defining those holes would be included here.
This annotation data is crucial for training your object detection model. The model will learn to recognize objects like "tea" based on the shapes (polygons) defined by the coordinates in the annotations.

In [5]:
import pandas as pd
import IPython
print(pd.__version__)
print(IPython.__version__)
!pip install tensorflow
!pip install tensorflow-object-detection-api
!pip install --upgrade dataset-tools


2.2.2
7.34.0
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 577.4/577.4 kB 11.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.5/54.5 kB 3.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.6/11.6 MB 57.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 117.2/117.2 kB 8.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 587.4/587.4 kB 32.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.4/76.4 kB 5.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.0/78.0 kB 5.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 69.1/69.1 kB 4.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 383.6/383.6 kB 22.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.5/133.5 kB 8.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 59.7/59.7 kB 4.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 769.2/769.2 kB 32.1 MB/s eta

In [1]:
!rm -rf models

In [2]:
!git clone https://github.com/tensorflow/models.git

Cloning into 'models'...
remote: Enumerating objects: 98333, done.
remote: Counting objects: 100% (839/839), done.
remote: Compressing objects: 100% (456/456), done.
remote: Total 98333 (delta 444), reused 696 (delta 362), pack-reused 97494 (from 1)
Receiving objects: 100% (98333/98333), 621.95 MiB | 24.68 MiB/s, done.
Resolving deltas: 100% (71428/71428), done.


In [3]:
!export PYTHONPATH=$PYTHONPATH:/path/to/models

In [4]:
import os
os.environ['PYTHONPATH'] += ":/content/models"

In [5]:
!pip install --user -r /content/models/official/requirements.txt

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 2.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.8/51.8 kB 2.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 242.5/242.5 kB 9.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 104.0/104.0 kB 7.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 51.8 MB/s eta 0:00:00
  Created wheel for seqeval: filename=seqeval-1.2.2-py3-none-any.whl size=16161 sha256=41cc82c51db380db622b3a0a12106b07443517e77e24a50bd1cfbb0d4342ed30
  Stored in directory: /root/.cache/pip/wheels/1a/67/4a/ad4082dd7dfc30f2abfe4d80a2ed5926a506eb8a972b4767fa
Successfully built seqeval
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


In [6]:
!apt-get install protobuf-compiler


Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
protobuf-compiler is already the newest version (3.12.4-1ubuntu7.22.04.1).
0 upgraded, 0 newly installed, 0 to remove and 49 not upgraded.


In [7]:
%cd models/research
!protoc object_detection/protos/*.proto --python_out=.


/content/models/research


In [8]:
import sys
sys.path.append('..')
sys.path.append('..models/research')
sys.path.append('..models/research/slim')


In [9]:
!pip install tf-slim

In [10]:
import tf_slim as slim

2024-10-26 06:21:20.069065: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-10-26 06:21:20.094874: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-10-26 06:21:20.102267: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-10-26 06:21:20.123713: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-10-26 06:21:22.288605: W tensorflow/compiler/tf2

In [11]:
!python object_detection/builders/model_builder_tf2_test.py


2024-10-26 06:21:36.935030: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-10-26 06:21:36.972117: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-10-26 06:21:36.984374: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-10-26 06:21:39.176314: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
Traceback (most recent call last):
  File "/content/models/research/object_detection/builders/model_builder_tf2_test.py", line 24, in <module>
    from object_detection.builders import model_builder
  File "/usr/local/lib/python3.10/dist-pac